replica/replica_test.go

```diff
@@ -1,11 +1,312 @@
 package replica_test
 
-// import (
-// 	. "github.com/onsi/ginkgo"
-// 	. "github.com/onsi/gomega"
-// 	. "github.com/renproject/hyperdrive/replica"
-// )
+import (
+	"fmt"
+	"reflect"
+	"sync"
 
-// var _ = Describe("Replica", func() {
+	"github.com/renproject/hyperdrive/block"
+	"github.com/renproject/hyperdrive/consensus"
+	"github.com/renproject/hyperdrive/shard"
+	"github.com/renproject/hyperdrive/sig"
+	"github.com/renproject/hyperdrive/sig/ecdsa"
+	"github.com/renproject/hyperdrive/tx"
 
-// })
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	. "github.com/renproject/hyperdrive/replica"
+)
+
+var _ = Describe("Replica", func() {
+
+	BeforeSuite(func() {
+		signer, err := ecdsa.NewFromRandom()
+		Expect(err).ShouldNot(HaveOccurred())
+		stateMachine := consensus.NewStateMachine(block.PolkaBuilder{}, block.CommitBuilder{}, 1)
+		transitionBuffer = newMockTransitionBuffer()
+
+		shard := shard.Shard{
+			Hash:        sig.Hash{},
+			BlockHeader: sig.Hash{},
+			BlockHeight: 0,
+			Signatories: sig.Signatories{signer.Signatory()},
+		}
+
+		pool = NewMockLifoPool()
+
+		dispatcher = newMockDispatcher()
+		replica = New(dispatcher, signer, pool, consensus.WaitForPropose(0, 0), stateMachine, transitionBuffer, block.NewBlockchain(), shard)
+
+	})
+
+	Context("when a new Transaction is sent using Transact", func() {
+
+		It("should update the TxPool", func() {
+			replica.Transact(tx.Transaction{})
+			Expect(pool.Length()).Should(Equal(1))
+		})
+
+	})
+
+	Context("when new Transitions are sent", func() {
+
+		testCases := generateTestCases()
+		for _, t := range testCases {
+			t := t
+
+			Context(fmt.Sprintf("when the replica gets transition - %s", reflect.TypeOf(t.inputTransition).Name()), func() {
+				It("should ", func() {
+
+					replica.Transition(t.inputTransition)
+					t.validateResults()
+				})
+			})
+		}
+
+	})
+})
+
+type TestCase struct {
+	inputTransition consensus.Transition
```

Comment:
Each test case should be sequence of `Transitions` sent to the `Replica`, as well as a starting `State` and expected final `State`. Then, for each `TestCase`, a fresh `Replica` should be creating and set to the starting `State`, run through the sequence of `Transitions`, and then checked to ensure that it has arrived at in the expected final `State`.

Individual transitions are not enough overall to verify the correctness of the `Replica` under many different scenarios, and re-using the same `Replica` across `TestCases` means that we lose the ability to run tests in parallel or cleanly isolated various different complex scenarios.

Reply:
Fixed in https://github.com/renproject/hyperdrive/pull/5